remote.1
.TH remote 1 "Dec 20, 2018"

.
.
.
.SH NAME
remote \- read infrared signals from remote controls via an audio adapter

.
.
.
.SH SYNOPSIS
.TP 7
.B remote
[\fI-f\fP] [\fI-c\fP] [\fI-l\fP] [\fI-d n\fP]
(\fIfile\fP|\fIaudio_device\fP) --
[\fIamplify_factor\fP [\fItrigger_bound\fP]]

.
.
.
.SH DESCRIPTION

Read codes from an infrared remote control via an adapter connected to the
microphone input of a sound card. Currently only the NEC, SHARP, SONY20 and
RC5 protocols are supported.

.
.
.
.SH OPTIONS

.TP
.B -f
input is a sequence of numbers in ascii, one per line, instead of an AU file
.TP
.B -c
accept the signals produced by \fBirblast\fP(\fI1\fP) via a loopback audio
device
.TP
.B -l
log input to file \fIlog.au\fP; the log file is in ascii and is called
\fIlog.txt\fP if also \fI-f\fP is given
.TP
.BI -d " n
debug protocol \fIn\fP; see \fIPROTOCOLS\fP, below
.TP
.B amplify_factor
-1 to invert, default 1
.TP
.B trigger_bound
the default is to use the background noise canceler, which requires a bit of
signal absence at the beginning and derives the bound from it
.TP
.B file|audio_device
input is read from file if it exists, otherwise from audio capture
device audio_device (list available: arecord -L)

.
.
.
.SH DETAILS

The input file is by default an AU file with header, 16 bit signed integer PCM.
This is also the format of the log file, which is just a copy of the input
saved to file. If requested by option \fI-f\fP, both are sequences of integers
in ascii, one per line.

Decoding the input signal requires knowing the approximate level of the noise
intensity (to be precise, the maximal value that is output by the chain of
signal filters when no actual signal is present). This level can be determined
in two ways. The default is to derive it from the initial part of the signal;
this assumes that no signal is present when the program is started.
Alternatively, it is passed as the \fItrigger_bound\fP argument; this lifts the
assumption that no signal is initially present, but requires the value of the
additional argument. A compromise is to use the default once for a given
combination of sound card and adapter, take notice of the background bounds and
then pass their maximal absolute value multiplied by two as the trigger bound.

.
.
.
.SH PROTOCOLS

The numbers of the protocols are as follows. Each protocol is implemented
twice: once for the direct signal and once for the inverted signal. Also, the
nec repeat codes are considered a different protocol than the codes for the
keys.

.TS
l l
---
l l.
number	protocol
1	nec
2	nec, inverted
3	nec repeat
4	nec repeat, inverted
5	nec2
6	nec2, inverted
7	nec2 repeat
8	nec2 repeat, inverted
9	sharp
10	sharp, inverted
11	sony12
12	sony12, inverted
13	sony20
14	sony20, inverted
15	rc5
16	rc5, inverted
.TE

.
.
.
.SH SEE ALSO

\fIlayout\fP(1), \fIsignal2pbm\fP(1)



remote.c
/*
 * remote.c
 *
 * Copyright (C) 2019 <sgerwk@aol.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/*
 * parse audio data as a remote protocol
 *
 * remote [-f] [-l] [-i] [-d n] (file|dev) -- [amplify_factor [trigger_bound]]
 *	-f	input is a sequence of numbers in ascii, one per line,
 *		instead of an AU file
 *	-c	allow receiving the output of irblast
 *	-l	log input to log.au or log.txt
 *	-d n	debug protocol n, from 1 to 14 so far
 *	amplify_factor
 *		-1 to invert, default 1
 *	trigger_bound
 *		the default is to use the background noise canceler,
 *		which requires a bit of signal absence at the beginning
 *		and derives the bound from it
 *	file|dev
 *		input is read from file if it exists, otherwise from audio
 *		device dev (list available: arecord -L)
 */

#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <inttypes.h>
#include "microphone.h"
#include "filters.h"
#include "protocols.h"

/*
 * filter debugging: stop the chain of filters at some point and print
 */
#define STOPHERE				\
	printf("%d\n", value);			\
	if (status.flush)			\
		fflush(stdout);			\
	continue;

/*
 * main
 */
int main(int argc, char *argv[]) {
	int opt;
	char *filename, *logfile = NULL;
	int debug, ascii, valleyfilter;
	int bound;
	double factor;
	struct status status;
	void *read, *microphone, *log;
	void *valley, *diff, *amplify, *maximal, *stabilize;
	void *background, *trigger, *runlength;
	int value;
	struct protocols_status *protocols_status;
	struct key *key;

					/* arguments */

	ascii = 0;
	valleyfilter = 0;
	debug = 0;
	while (-1 != (opt = getopt(argc, argv, "fcld:")))
		switch (opt) {
		case 'l':
			logfile = "log.au";
			break;
		case 'f':
			ascii = 1;
			break;
		case 'c':
			valleyfilter = 1;
			break;
		case 'd':
			debug = atoi(optarg);
			break;
		}
	if (ascii && logfile)
		logfile = "log.txt";

	argc -= optind - 1;
	argv += optind - 1;
	if (argc - 1 < 1)
		filename = "default";
	else
		filename = argv[1];
	factor = argc - 1 >= 2 ? atof(argv[2]) : 1;
	bound = argc - 1 >= 3 ? atoi(argv[3]) : -1;

					/* init filters and protocols */

	read =             read_init(filename, ascii, &status);
	if (read != NULL)
		microphone = NULL;
	else {
		microphone = microphone_init(filename, &status);
		if (microphone == NULL) {
			printf("cannot open input file\n");
			exit(EXIT_FAILURE);
		}
	}
	log =               log_init(logfile, ascii, &status);
	valley =         valley_init(10, &status);
	diff =             diff_init(&status);
	amplify =       amplify_init(factor, &status);
	maximal =       maximal_init(11, &status);
	stabilize =   stabilize_init(&status);
	trigger =       trigger_init(bound, &status);
	background = background_init(&status);
	runlength =   runlength_init(&status);

	protocols_status = protocols_init(debug);
	
					/* process values */

	while (! status.ended) {

		// filter testing: STOPHERE to cut the pipe of filters short

		value = 0;
		if (read)
			FILTER_VALUE(read, value, read, &status)
		if (microphone)
			FILTER_VALUE(microphone, value, microphone, &status)
		FILTER_VALUE(log, value, log, &status)
		if (valleyfilter)
			FILTER_VALUE(valley, value, valley, &status);
		FILTER_VALUE(diff, value, diff, &status)
		FILTER_VALUE(amplify, value, amplify, &status)
		FILTER_VALUE(stabilize, value, stabilize, &status)
		FILTER_VALUE(maximal, value, maximal, &status)
		if (bound == -1)
			FILTER_VALUE(background, value, background, &status)
		else
			FILTER_VALUE(trigger, value, trigger, &status)
		FILTER_VALUE(runlength, value, runlength, &status)

		if (! debug) {
			printf("*");
			fflush(stdout);
		}

		key = protocols_value(value, protocols_status);
		if (key) {
			printf("\n");
			printkey(key);
			printf("\n");
		}
	}

					/* finish filters */

	if (read)
		read_end(read, &status);
	if (microphone)
		microphone_end(microphone, &status);
	log_end(log, &status);
	valley_end(valley, &status);
	diff_end(diff, &status);
	amplify_end(amplify, &status);
	stabilize_end(stabilize, &status);
	maximal_end(maximal, &status);
	trigger_end(trigger, &status);
	background_end(background, &status);
	value = runlength_end(runlength, &status);
	protocols_value(value, protocols_status);
	protocols_end(protocols_status);

	if (! debug)
		printf("\n");

	return EXIT_SUCCESS;
}

